ambry-router/src/main/java/com/github/ambry/router/PutOperation.java

```diff
@@ -1211,22 +1229,33 @@ void onFillComplete(boolean updateMetric) {
     }
 
     /**
-     * Fill the buffer of the current chunk with the data from the given {@link ByteBuffer}.
-     * @param channelReadBuffer the {@link ByteBuffer} from which to read data.
+     * Fill the buffer of the current chunk with the data from the given {@link ByteBuf}.
+     * @param channelReadBuf the {@link ByteBuf} from which to read data.
      * @return the number of bytes transferred in this operation.
      */
-    int fillFrom(ByteBuffer channelReadBuffer) {
-      int toWrite = Math.min(channelReadBuffer.remaining(), buf.remaining());
-      if (channelReadBuffer.remaining() > buf.remaining()) {
-        // Manipulate limit of the source buffer in order to read only enough to fill the chunk
-        int savedLimit = channelReadBuffer.limit();
-        channelReadBuffer.limit(channelReadBuffer.position() + buf.remaining());
-        buf.put(channelReadBuffer);
-        channelReadBuffer.limit(savedLimit);
+    int fillFrom(ByteBuf channelReadBuf) {
+      int toWrite;
+      if (buf == null) {
+        // If current buf is null, then only read the up to routerMaxPutChunkSizeBytes.
+        toWrite = Math.min(channelReadBuf.readableBytes(), routerConfig.routerMaxPutChunkSizeBytes);
+        buf = channelReadBuf.readRetainedSlice(toWrite);
       } else {
-        buf.put(channelReadBuffer);
+        int remainingSize = routerConfig.routerMaxPutChunkSizeBytes - buf.readableBytes();
+        toWrite = Math.min(channelReadBuf.readableBytes(), remainingSize);
+        ByteBuf remainingSlice = channelReadBuf.readRetainedSlice(toWrite);
+        // buf already has some bytes
```

Comment:
Can you explain a little bit how fillFrom triggered?  I am not familiar with this area.

For 1 MB blob put, will it trigger fillFrom multiple times? 

My concern here is if this will cause many small ByteBufs in `buf`.

Reply:
It depends on how the http content works. For http put request, frontend would receive a bunch of netty httpContents, each http content carries a netty bytebuf, which is `channelReadBuf`. PutManager has a running thread that calls fillFrom periodically.

Reply:
If we have a 1MB blob, if the frontend receives 10 100K bytebuf, then the final `buf` would have 10 small ByteBuf. But the important thing is this is zero-copy bytebuf. We don't have to move any bytes from one place to another.